README.md
<img height="60" src="https://raw.githubusercontent.com/urpflanze-org/core/master/docs/assets/images/logo-for-github.svg">

## Synopsis

This package is the core used by the [Urpflanze](https://github.com/urpflanze-org/urpflanze) javascript library to generate the scene.

It deals with creating two-dimensional shapes, repeating them, manipulating them point by point and encapsulating them.

You can use it in the browser or in node.

## Motivations

The creation of this library comes from the need to create simple APIs for
manage the repetition of primitive shapes and the possibility of applying transformations to each of them, applying transformations on the points avoiding the use of canvas transformations.

Another need - which then became one of the main features - was to be able to encapsulate the result of a generation and manage it as if it were a new shape.

## Donate

I am trying to create a tool for those who want to approach the world of programming
or for programmers who want to approach the world of creative coding.

I have spent a lot of time and will spend more to support this project.
I also have in mind a **[web editor](https://github.com/urpflanze-org/editor)** (open-source) where you can use the features of this library in the browser.

You can see a preview [here](https://editor.urpflanze.org)

[![](https://img.shields.io/badge/donate-paypal-003087.svg?logo=paypal)](https://www.paypal.me/genbs)
[![](https://img.shields.io/badge/donate-ko--fi-29abe0.svg?logo=ko-fi)](https://ko-fi.com/urpflanze)

[![](https://img.shields.io/badge/bitcoin-1CSQq4aMmsA71twvyZHZCjmeB2AmQGCPNq-f7931a.svg?logo=bitcoin)](https://explorer.btc.com/btc/address/1CSQq4aMmsA71twvyZHZCjmeB2AmQGCPNq)
[![](https://img.shields.io/badge/ethereum-0x9086c4bb7015c1d6dc79162d02e7e1239c982c01-ecf0f1.svg?logo=ethereum)](https://etherscan.io/address/0x9086c4bb7015c1d6dc79162d02e7e1239c982c01)

---

## Menu

- [Synopsis](#synopsis)
- [Motivations](#motivations)
- [Donate](#donate)
- [Menu](#menu)
- [Installation](#installation)
- [Creating a shape](#creating-a-shape)
	- [ShapeBuffer](#shapebuffer)
	- [ShapeLoop](#shapeloop)
- [Primitive shapes](#primitive-shapes)
		- [ShapeBuffer](#shapebuffer-1)
		- [ShapeLoop](#shapeloop-1)
	- [Modifiers](#modifiers)
- [Repetitions](#repetitions)
	- [Ring repetitions](#ring-repetitions)
	- [Matrix repetitions](#matrix-repetitions)
- [Manage repetitions](#manage-repetitions)
	- [Repetitions examples](#repetitions-examples)
	- [List of properties](#list-of-properties)
- [Encapsulation](#encapsulation)
	- [Shape](#shape)
	- [Group](#group)
	- [Using repetition property of the encapsulator](#using-repetition-property-of-the-encapsulator)
- [Recursion](#recursion)
- [ShapeFollow](#shapefollow)
- [Vertex Callback](#vertex-callback)
- [Scene](#scene)
	- [Simple Drawer](#simple-drawer)
- [Examples](#examples)

---

## Installation

You can install the library with the command:

```bash
npm i @urpflanze/core --save
```

At the end you can import Urpflanze into your project

```javascript
/**
 * Full importing
 */
import * as Urpflanze from '@urpflanze/core'

const scene = new Urpflanze.Scene()

/**
 * Selective import
 */
import { Scene } from '@urpflanze/core'

const scene = new Scene()
```

---

## Creating a shape

### ShapeBuffer

The _ShapeBuffer_ is the shape to which you can pass a buffer of points.
It accepts the `shape` property which is an array of points _[x0, y0, x1, y1, ..., xn, yn]_.

The array of points will be adapted between a range of -1 and 1.

Example:

```javascript
import { ShapeBuffer } from '@urpflanze/core'

const rect = new ShapeBuffer({
	shape: [-1, -1, 1, -1, 1, 1, -1, -1],
	sideLength: [10, 10],
})

rect.generate() // Apply properties

console.log(rect.getBuffer())

// Output:
// Float32Array(8) [
//   -10, -10,  10, -10,
//    10,  10, -10, -10
// ]
```

### ShapeLoop

The _ShapeLoop_ is a shape generated by a loop, it is recommended to return values ​​between a range of -1 and 1

```javascript
import { ShapeLoop } from '@urpflanze/core'

const circle = new Urpflanze.ShapeLoop({
	sideLength: [10, 10],
	loop: {
		start: 0,
		end: Math.PI * 2,
		inc: (Math.PI * 2) / 100, // (end - start) / 100 for generate 100 points
		vertex: shapeLoopRepetition => [
			// shapeLoopRepetition.current start from 0 and end to 2 PI,
			// so you can use it as a angle
			Math.cos(shapeLoopRepetition.current),
			Math.sin(shapeLoopRepetition.current),
		],
	},
	bClosed: true, // flag to determinate the shape is closed
})

circle.generate()

console.log(circle.getBuffer().length)

// Output:
// 200 / 2 = 100 points
```

---

## Primitive shapes

In this package there are already some primitive shapes.
You can use the static `getBuffer` method to receive a buffer.

```javascript
console.log(Urpflanze.Rect.getBuffer(/*{ ...props }*/))
// Float32Array(8) [
//    1,  1, -1,  1,
//   -1, -1,  1, -1
// ]
```

List of all primitives:

#### ShapeBuffer

[`Line`](https://docs.urpflanze.org/core/#/ref/Line)
[`Triangle`](https://docs.urpflanze.org/core/#/ref/Triangle)
[`Rect`](https://docs.urpflanze.org/core/#/ref/Rect)

#### ShapeLoop

[`Polygon`](https://docs.urpflanze.org/core/#/ref/Polygon)
[`Circle`](https://docs.urpflanze.org/core/#/ref/Circle)
[`Lissajous`](https://docs.urpflanze.org/core/#/ref/Lissajous)
[`Spiral`](https://docs.urpflanze.org/core/#/ref/Spiral)
[`Rose`](https://docs.urpflanze.org/core/#/ref/Rose)
[`SuperShape`](https://docs.urpflanze.org/core/#/ref/SuperShape)

---

### Modifiers

On primitive shapes you can apply modifiers to the buffer before transformations are applied:

```javascript
const squircle = new Urpflanze.Rect({
	modifiers: [new Urpflanze.Modifiers.Smooth({ level: 3, tension: 0.3 })],
})
```

![](https://docs.urpflanze.org/core/assets/images/readme/modifiers-1.png)

The list of modifiers is:

- `Adapt`: Fit the shape into a rectangle
- `Close`: Adds a close point to the buffer
- `Mirror`: clones the shape on the x, y axes
- `Offset`: Slice of a buffer
- `Smooth`: Makes the shape smoother
- `Subdivide`: Adds midpoints on the edges of the shape
- `Solidify`: (in progress) Tread the edge of the shape

---

## Repetitions

Using Urpflanze you can manage two types of repetitions: **ring** or **matrix**.

### Ring repetitions

For this type of repetition you can set a numeric value to the `repetitions` property to indicate the number of times it will repeat and the` distance` property to indicate the distance from the center.

```javascript
const rects = new Urpflanze.Rect({
	repetitions: 8,
	distance: 100,
	sideLength: 25,
})
```

![](https://docs.urpflanze.org/core/assets/images/readme/repetition-ring-1.png)

Basically the shapes will be rotated towards the center, if you want to avoid this effect you have to rotate the vorma inversely to the current angle of the repetition.

```javascript
const rects = new Urpflanze.Rect({
	repetitions: 8,
	sideLength: 25,
	distance: 100,
	rotateZ: ({ repetition }) => -repetition.angle,
})
```

![](https://docs.urpflanze.org/core/assets/images/readme/repetition-ring-2.png)

### Matrix repetitions

To repeat the shape as an array, just pass an Array of numbers indicating the number of rows and columns to the `repetitions` property. The `distance` property in this case will also be an Array containing the distance between the rows and columns.

```javascript
const matrix = new Urpflanze.Rect({
	repetitions: [3, 4],
	sideLength: 20,
	distance: [80, 50],
})
```

![](https://docs.urpflanze.org/core/assets/images/readme/repetition-matrix-1.png)

---

## Manage repetitions

To manage the repetitions you can pass a function to the properties instead of a constant.

The argument of the function which is of type [IPropArguments](https://docs.urpflanze.org/urpflanze/#/ref/IPropArguments).

Inside it we find the `repetition` property which - like any object that implements a [IBaseRepetition](https://docs.urpflanze.org/urpflanze/#/ref/IBaseRepetition) - contains the following properties:

- `index` the current index, from 1 to _count_
- `count` the total number of repetitions
- `offset` an index ranging from 0 to 1 which does not depend on the number of repetitions. For example, if the number of repetitions is 3, the offset value will be 0 - 0.5 - 1

For matrix repetitions you can also use `repetition.row` and` repetition.col` also of type [IBaseRepetition](https://docs.urpflanze.org/urpflanze/#/ref/IBaseRepetition)

### Repetitions examples

```javascript
const spiral = new Urpflanze.Rect({
	repetitions: 8,
	sideLength: 25,
	distance: ({ repetition }) => repetition.offset * 100,
	scale: ({ repetition }) => repetition.offset,
})
```

![](https://docs.urpflanze.org/core/assets/images/readme/repetition-1.png)

```javascript
const matrix = new Urpflanze.Rect({
	repetitions: [4],
	sideLength: 25,
	distance: 50,
	scale: ({ repetition }) => {
		// [0, 0] is center of repetition, you can set value between [-1, -1] (left - top angle) and [1, 1] (right - bottom angle)
		return Urpflanze.distanceFromRepetition(repetition, [0, 0]),
	}
})
```

![](https://docs.urpflanze.org/core/assets/images/readme/repetition-2.png)

### [List of properties](https://docs.urpflanze.org/core/#/ref/IShapeBaseSettings)

---

## Encapsulation

To be able to encapsulate a shape you can use the `Shape` class to which you can pass the property
`shape` which is a` ShapePrimitive` (ShapeBuffer or ShapeLoop) or a `Group`.

### Shape

```javascript
const lines = new Urpflanze.Line({
	repetitions: 20,
	sideLength: 25,
	distance: 50,
})

const container = new Urpflanze.Shape({
	shape: lines,

	repetitions: [3], // [3, 3]
	distance: 100,
	scale: 0.5, // scale all repetitions of lines
})

const final = new Urpflanze.Shape({
	shape: container,

	repetitions: 6,
	distance: 120,
	scale: 0.4,
	perspective: 0.99,
	rotateY: Urpflanze.toRadians(60),
})
```

|                                     lines                                     |                                   container                                   |                                     final                                     |
| :---------------------------------------------------------------------------: | :---------------------------------------------------------------------------: | :---------------------------------------------------------------------------: |
| ![](https://docs.urpflanze.org/core/assets/images/readme/encapsulation-1.png) | ![](https://docs.urpflanze.org/core/assets/images/readme/encapsulation-2.png) | ![](https://docs.urpflanze.org/core/assets/images/readme/encapsulation-3.png) |

### Group

```javascript
const group = new Urpflanze.Group({
	repetitions: 4,
	sideLength: 15,
	distance: 25,
	rotateZ: Urpflanze.toRadians(45),
})

group.add(
	new Urpflanze.Circle(),
	new Urpflanze.Rect(),
	new Urpflanze.Line({
		rotateZ: 0,
	})
)

const shape = new Urpflanze.Shape({
	shape: group,
	repetitions: 8,
	distance: 100,
	rotateZ: ({ repetition }) => -repetition.angle,
})
```

|                                 group                                 |                                 shape                                 |
| :-------------------------------------------------------------------: | :-------------------------------------------------------------------: |
| ![](https://docs.urpflanze.org/core/assets/images/readme/group-1.png) | ![](https://docs.urpflanze.org/core/assets/images/readme/group-2.png) |

### Using repetition property of the encapsulator

You can use the `repetition` object of whoever encapsulates a shape by setting the `shapeUseParent` property.
This parameter is optional since a new buffer of points will be generated at each repetition of the encapsulator.

```javascript
const rects = new Urpflanze.Rect({
	repetitions: [5],
	sideLength: 10,
	distance: 20,
	scale: ({ repetition, parent }) => {
		return repetition.offset * parent.repetition.offset
	},
})

const container = new Urpflanze.Shape({
	shapeUseParent: true, // <--
	shape: rects,
	repetitions: [5],
	distance: 50,
	scale: 0.4,
})
```

![](https://docs.urpflanze.org/core/assets/images/readme/parent-rep-1.png)

---

## Recursion

Another possibility is to use the `ShapeRecursive` to repeat any `Shape` on each of its points.

You can use the `recursion` property of type [IRecursionRepetition](https://docs.urpflanze.org/core/#/ref/IRecursionRepetition)

```javascript
const rect = new Urpflanze.Rect({
	sideLength: 50,
})

const container = new Urpflanze.ShapeRecursive({
	// shapeUseRecursion: true,
	// [prop]: ({ recursion }) => ...
	shape: rect,
	recursionScale: 2,
	recursions: 4,
})
```

![](https://docs.urpflanze.org/core/assets/images/readme/recursion-1.png)

---

## ShapeFollow

With the `ShapeFollow` you can repeat the `shape` on each point of the `follow`

In this example we can also see the use of the interpolation function between buffers:

```javascript
/**
 * Repeat a Polygon on the buffer that morphs from Rect to Star
 *
 * for animation using https://github.com/urpflanze-org/animation
 * and https://github.com/urpflanze-org/drawer-canvas for draw and rendering
 */
const scene = new Urpflanze.Scene()

const [from, to] = Urpflanze.prepareBufferForInterpolation(
	Urpflanze.Rect.getBuffer({
		modifiers: [new Urpflanze.Modifiers.Smooth({ closed: false, tension: 0.3, level: 4 })],
	}),
	Urpflanze.Star.getBuffer({
		spikes: 6,
		innerRadius: 1.3,
		modifiers: [new Urpflanze.Modifiers.Smooth({ closed: false, tension: 0.7, level: 5 })],
	})
)

scene.add(
	new Urpflanze.ShapeFollow({
		shape: new Urpflanze.Polygon({
			sideNumber: 6,
			sideLength: 20,
			rotateZ: t => {
				const d = 2000
				const o = Math.cos(t.parent.follow.offset * Urpflanze.PI2) * 500 // set shapeUseFollor for bind follow IBaseRepetition

				const time = Animation.clock(scene.currentTime, d - o, true, 'normal', o)
				return Animation.Easings.quadraticOut(time, 0, Urpflanze.PI2, d - o)
			},
			drawer: {
				stroke: '#ffffff22',
			},
		}),
		follow: new Urpflanze.ShapeBuffer({
			sideLength: 100,
			shape: () => {
				// Interpolate Rect (from) and Star (to) with easing
				return Urpflanze.interpolate(
					from,
					to,
					Animation.Easings.elasticInOut(Animation.clockOffset(scene.currentTime, 2000), 0, 1, 1, 1.2)
				)
			},
		}),

		shapeUseFollow: true, // use follow repetition in Rect (for rotateZ)
	})
)

const drawer = new DrawerCanvas(scene, document.body, {}, 4000, 24)
drawer.startAnimation()
```

![](https://docs.urpflanze.org/core/assets/images/readme/shape-follow.gif)

---

## Vertex Callback

The vertexCallback property is a function that is called at each point of the shape of each repetition.

The function takes 3 arguments:

- `vertex`: [number, number] current vertex
- `vertexRepetition`: [IBaseRepetition](https://docs.urpflanze.org/urpflanze/#/ref/IBaseRepetition) for the vertices of the current repetition
- `propArguments`: [IPropArguments](https://docs.urpflanze.org/urpflanze/#/ref/IPropArguments)

```javascript
const rects = new Urpflanze.Rect({
	repetitions: [10, 1],
	sideLength: 100,
	scale: propArguments => propArguments.repetition.row.offset * 0.95 + 0.05,
	modifiers: new Urpflanze.Modifiers.Subdivide({ level: 5 }),
	vertexCallback: (vertex, vertexRepetition, propArguments) => {
		const angle = Math.atan2(vertex[1], vertex[0])

		const x = Math.cos(angle)
		const y = Math.sin(angle)

		const offset = propArguments.repetition.row.offset ** 2 * 20
		const noise = Urpflanze.noise('seed', vertexRepetition.offset * 10) * offset

		vertex[0] += x * noise
		vertex[1] += y * noise
	},
})
```

![](https://docs.urpflanze.org/core/assets/images/readme/vertexcallback-1.png)

---

## Scene

You can use the shapes independently or you can add them to a scene.
When a shape is added to the scene it will be arranged in the center of it, adding an offset to all points.

_Use without the scene:_

```javascript
const rect = new Urpflanze.Rect({ sideLength: 25 })
rect.generate()

console.log(rect.getBounding())
// Output:
//
// { cx: 0, cy: 0, x: -25, y: -25, width: 50, height: 50 }
// # left, top point: (-25, -25) | right, bottom point: (25, 25)
```

_Using the scene:_

```javascript
const scene = new Urpflanze.Scene({ width: 100, height: 100 })
const rect = new Urpflanze.Rect({ sideLength: 25 })
scene.add(rect)
scene.update()

console.log(rect.getBounding())
// Output:
//
// {
//	cx: 50, cy: 50, 	  # Center of scene
//	x: 25, y: 25,  		  # Center of scene - sideLength
//	width: 50, height: 50 # sideLength * 2
// }
// # left, top point: (25, 25) | right, bottom point: (75, 75)
```

### Simple Drawer

When you call the `generate()` method on a shape a buffer of type Array<[IBufferIndex](https://docs.urpflanze.org/core/#/ref/IBufferIndex)> is created containing the information on the current repetition of shape and the reference index of the total buffer (`getBuffer()`)

if repetitions are statica (and ShapeLoop has static loop) the IndexedBuffer will only generate once.

![](https://docs.urpflanze.org/core/assets/images/buffers.svg)

```javascript

// scene.add(...)

const time = Date.now()

scene.currentTime = time
const sceneChilds = scene.getChildren()
for(let i = 0, len = sceneChilds.length; i < len; i++) {
	// Generate Buffer and IndexedBuffer
	sceneChilds[i].generate(time, true)

	// Buffer of indexing (https://docs.urpflanze.org/core/#/ref/IBufferIndex)
	const childIndexedBuffer = sceneChilds[i].getIndexedBuffer()
	const childBuffer = sceneChilds[i].getBuffer()


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.5.12](https://github.com/urpflanze-org/core/compare/v0.5.11...v0.5.12) (2021-12-29)


### Features

* add isGenerated method ([2febdc3](https://github.com/urpflanze-org/core/commits/2febdc37a67e54867200d0f36e5c7fac32257f20))


### Bug Fixes

* change script order ([55ed4e3](https://github.com/urpflanze-org/core/commits/55ed4e399bb86af31df04636c3d4bf00e4cea816))

### [0.5.11](https://github.com/urpflanze-org/core/compare/v0.5.10...v0.5.11) (2021-11-11)


### Bug Fixes

* shallow buffer ([33e21d9](https://github.com/urpflanze-org/core/commits/33e21d9b6e9a522a32f3cd4ec4ef8e2d2592b662))

### [0.5.10](https://github.com/urpflanze-org/core/compare/v0.5.9...v0.5.10) (2021-11-01)

### [0.5.9](https://github.com/urpflanze-org/core/compare/v0.5.8...v0.5.9) (2021-08-23)


### Bug Fixes

* **utilities:** :bug: fix noise function ([6c9ac3a](https://github.com/urpflanze-org/core/commits/6c9ac3a959a9b5c1f9b0cac1482873a82d58b470))

### [0.5.8](https://github.com/urpflanze-org/core/compare/v0.5.7...v0.5.8) (2021-08-23)


### Features

* add loopMeta to ShapeLoop loop vertex callback ([414a2fd](https://github.com/urpflanze-org/core/commits/414a2fdcee146ee7017bd5b7c2784368c9cd8051))
* add noise min max value ([71d6523](https://github.com/urpflanze-org/core/commits/71d6523224772ba0c6e197b9c112a4a78740cfb4))


### Bug Fixes

* :bug: fix anchor when resize ([13579e0](https://github.com/urpflanze-org/core/commits/13579e0bc52051ab260cfdfa1de53c453029ea16))
* fix static getBuffer argument ([853cffb](https://github.com/urpflanze-org/core/commits/853cffb6134208cc3a99a4a2e246901f62a07068))

### [0.5.7](https://github.com/urpflanze-org/core/compare/v0.5.6...v0.5.7) (2021-07-29)

### 0.5.6 (2021-07-27)

- Fix drawer type

### 0.5.5 (2021-07-17)

- Add anchor to ShapeBase

### 0.5.4 (2021-07-17)

- Fix anchor settings
- Fix module path for cdn

### 0.5.3 (2021-07-17)

- Add "build" folder to files in package.json

### 0.5.2 (2021-07-17)

- Improve building system (ESM/CJS/Browser)

### 0.5.1 (2021-06-27)

- Fix absolute imports

### 0.5.0 (2021-06-25)

- Add ShapeFollow
- Fix ShapeLoop isStaticIndexed
- Add 'interpolate' utility for buffer morphing
- ShapeBuffer add dynamic shape
- Add Modifiers
- Add SingleRepetitionBounding in indexedBuffer

### 0.4.6 (2021-03-29)

- Fix typescript import path

### 0.4.5 (2021-03-24)

- Interface refactoring

### 0.4.4 (2021-03-18)

- Remove buffer adaption when generate ShapeLoop or ShapeBuffer

### 0.4.3 (2021-03-17)

- Change module exports

### 0.4.2 (2021-03-16)

- Change build target from CJS to ES6

### 0.4.1 (2021-03-01)

- Rename 'style' to 'drawer' property for shape primitive

### 0.4.0 (2021-03-05)

- Adapt buffer moved in shape buffer
- Rename 'angle' to 'current' into IShapeLoopRepetition

### 0.3.3 (2021-02-24)

- Fix recursion on not primitive shape

### 0.3.2 (2021-02-11)

- Fix getBufferLength
- Fix clearBuffer for ShapeRecursive
- Add ShapeLoop subdivide for static inc

### 0.3.0 (2021-01-25)

- Remove time and Context from IPropArguments

### 0.2.4 (2021-01-24)

- Minor fix

### 0.2.3 (2021-01-09)

- Fix parent when indexing

### 0.2.2 (2020-12-14)

- Add RecoursiveShape
- Move to 'style' the properties used by drawer
- Rename RegularPolygon to Polygon

### 0.1.0-alpha.2 (2020-11-12)

- In the generation of the shapes the 'squeeze' was moved before apply the transformations


src/Utilities.ts
import * as SimplexNoise from 'simplex-noise'

import { ERepetitionType, IRepetition } from './types/repetitions'

import Vec2 from './math/Vec2'

const measurement = typeof performance !== 'undefined' ? performance : Date
/**
 * Get current timestamp in milliseconds
 *
 * @category Utilities
 * @returns {number}
 */
export function now(): number {
	return measurement.now()
}

// aOr: (...args: Array<any>): any => {
// 	for (let i = 0; i < args.length; i++) if (Utilities.isDef(args[i])) return args[i]
// },

/**
 * Convert number from radians to degrees
 *
 * @category Utilities
 *
 * @example
 * ```javascript
 * Urpflanze.toDegrees(Math.PI) // 180
 * ```
 *
 * @param {number} radians
 * @returns {number}
 */
export function toDegrees(radians: number): number {
	return (radians * 180) / Math.PI
}

/**
 * Convert angle from degrees to radians
 * @example
 * ```javascript
 * Urpflanze.toRadians(180) // 3.141592653589793
 * ```
 *
 * @category Utilities
 * @param {number} degrees
 * @returns {number}
 */
export function toRadians(degrees: number): number {
	return (degrees * Math.PI) / 180
}

/**
 * Linear interpolation from `a` when `i` as 0 an `b` when `i' as 1
 *
 * @category Utilities
 * @param {number} a
 * @param {number} b
 * @param {number} i
 * @returns {number}
 */
export function lerp(a: number, b: number, i: number): number {
	return (1 - i) * a + i * b
}

/**
 * Return number between min and max
 *
 * @category Utilities
 * @example
 * ```javascript
 * Urpflanze.clamp(0, 1, 1.2) // 1
 * Urpflanze.clamp(0, 1, -2) // 0
 * ```
 * @param {number} min
 * @param {number} max
 * @param {number} value
 * @returns {number}
 */
export function clamp(min: number, max: number, value: number): number {
	return value <= min ? min : value >= max ? max : value
}

/**
 * Map number between refMin e refMax from min and max
 *
 * @category Utilities
 *
 * @example
 * ```javascript
 * Urpflanze.relativeClamp(0, 1, 0.5, 100, 200) // 150
 * ```
 *
 * @param {number} refMin
 * @param {number} refMax
 * @param {number} value
 * @param {number} toMin
 * @param {number} toMax
 * @returns {number}
 */
export function relativeClamp(refMin: number, refMax: number, value: number, toMin: number, toMax: number): number {
	return clamp(toMin, toMax, ((value - refMin) / (refMax - refMin)) * (toMax - toMin) + toMin)
}

/**
 * @internal
 * @ignore
 */
const noises: {
	[key: string]: SimplexNoise
} = {
	random: new SimplexNoise(Math.random),
}

/**
 * <a href="https://github.com/jwagner/simplex-noise.js" target="_blank">SimplexNoise</a>
 * Use 'random' as seed property for random seed.
 * Return value between -1 and 1
 *
 * @category Utilities
 *
 * @param {string} [seed='random']
 * @param {number} [x=0]
 * @param {number} [y=0]
 * @param {number} [z=0]
 * @param {number} [min=-1]
 * @param {number} [max=-1]
 * @returns {number} between -1 and 1
 */
export function noise(seed = 'random', x = 0, y = 0, z = 0, min = -1, max = 1): number {
	if (typeof noises[seed] === 'undefined') {
		noises[seed] = new SimplexNoise(seed)
	}
	const value = noises[seed].noise3D(x, y, z)

	return min !== -1 || max !== 1 ? (0.5 + value * 0.5) * (max - min) + min : value
}

/**
 * @internal
 * @ignore
 */
const randoms: {
	[key: string]: () => number
} = {}

/**
 * Random number generator
 * @example
 * ```javascript
 * 	Urpflanze.random('seed') // 0.9367527104914188
 * ```
 *
 * @category Utilities
 * @param {string} seed
 * @param {number} min
 * @param {number} max
 * @param {number} decimals
 * @returns {number}
 */
export function random(seed?: string | number | null, min = 0, max = 1, decimals?: number): number {
	const key: string = seed + ''

	if (typeof randoms[key] === 'undefined') {
		const seed = xmur3(key)
		randoms[key] = sfc32(seed(), seed(), seed(), seed())
	}

	const value = min + randoms[key]() * (max - min)

	return typeof decimals !== 'undefined' ? Math.round(value * 10 ** decimals) / 10 ** decimals : value
}

/**
 *
 * @internal
 * @param str
 * @returns
 */
function xmur3(str: string) {
	let i = 0,
		h = 1779033703 ^ str.length
	for (; i < str.length; i++) (h = Math.imul(h ^ str.charCodeAt(i), 3432918353)), (h = (h << 13) | (h >>> 19))

	return function () {
		h = Math.imul(h ^ (h >>> 16), 2246822507)
		h = Math.imul(h ^ (h >>> 13), 3266489909)
		return (h ^= h >>> 16) >>> 0
	}
}

/**
 * @internal
 * @param a
 * @param b
 * @param c
 * @param d
 * @returns
 */
function sfc32(a: number, b: number, c: number, d: number) {
	return function () {
		a >>>= 0
		b >>>= 0
		c >>>= 0
		d >>>= 0
		let t = (a + b) | 0
		a = b ^ (b >>> 9)
		b = (c + (c << 3)) | 0
		c = (c << 21) | (c >>> 11)
		d = (d + 1) | 0
		t = (t + d) | 0
		c = (c + t) | 0
		return (t >>> 0) / 4294967296
	}
}

/**
 * Return angle (atan) from offset (or center) for matrix repetition.
 * Offset is array between [-1, -1] and [1, 1].
 * The return value is between -Math.PI / 2 and Math.PI / 2
 *
 * @category Utilities
 *
 * @param {IRepetition} repetition
 * @param {[number, number]} offsetFromCenter
 * @returns {number} between -Math.PI / 2 and Math.PI / 2
 */
export function angleFromRepetition(repetition: IRepetition, offsetFromCenter: [number, number] = [0, 0]): number {
	if (repetition.type === ERepetitionType.Matrix) {
		const centerMatrix = [(repetition.col.count - 1) / 2, (repetition.row.count - 1) / 2]

		centerMatrix[0] += centerMatrix[0] * offsetFromCenter[0]
		centerMatrix[1] += centerMatrix[1] * offsetFromCenter[1]

		const x = repetition.col.index - 1 - centerMatrix[0]
		const y = repetition.row.index - 1 - centerMatrix[1]

		return x === 0 ? 0 : Math.atan(y / x)
	}

	return (repetition.angle - Math.PI) / 2
}

/**
 * Return angle (atan2, 4 quadrants) from offset (or center) for matrix repetition.
 * Offset is array between [-1, -1] and [1, 1].
 * The return value is between -Math.PI an Math.PI
 *
 * @category Utilities
 *
 * @param {IRepetition} repetition
 * @param {[number, number]} offsetFromCenter
 * @returns {number} between -Math.PI an Math.PI
 */
export function angle2FromRepetition(repetition: IRepetition, offsetFromCenter: [number, number] = [0, 0]): number {
	if (repetition.type === ERepetitionType.Matrix) {
		const centerMatrix = [(repetition.col.count - 1) / 2, (repetition.row.count - 1) / 2]

		centerMatrix[0] += centerMatrix[0] * offsetFromCenter[0]
		centerMatrix[1] += centerMatrix[1] * offsetFromCenter[1]

		const x = repetition.col.index - 1 - centerMatrix[0]
		const y = repetition.row.index - 1 - centerMatrix[1]

		return x === 0 ? 0 : Math.atan2(y, x)
	}

	return repetition.angle - Math.PI
}

/**
 * Return distance from offset (or center) for matrix repetition.
 * The return value is between 0 and 1
 *
 * @category Utilities
 *
 * @param {IRepetition} repetition
 * @param {[number, number]} offsetFromCenter offset relative to distance prop
 * @returns {number} between 0 and 1
 */
export function distanceFromRepetition(repetition: IRepetition, offsetFromCenter: [number, number] = [0, 0]): number {
	if (repetition.type === ERepetitionType.Matrix) {
		const centerMatrix = [0.5, 0.5]

		centerMatrix[0] += centerMatrix[0] * offsetFromCenter[0]
		centerMatrix[1] += centerMatrix[1] * offsetFromCenter[1]

		const current = [repetition.col.offset, repetition.row.offset]

		return Vec2.distance(current, centerMatrix)
	}

	return 1
}

/// Interpolation

/**
 * Evenly distributes a number of points in a buffer
 *
 * @category Utilities.Buffer interpolation
 * @export
 * @param {Float32Array} buffer current buffer
 * @param {number} pointsToAdd points to add
 * @return {*}  {Float32Array}
 */
export function distributePointsInBuffer(buffer: Float32Array, pointsToAdd: number): Float32Array {
	const bufferLen = buffer.length
	const pointsLen = bufferLen / 2
	const finalBufferLength = (pointsLen + pointsToAdd) * 2
	const edges = pointsLen - 1

	if (edges > 1) {
		const lastPoint = bufferLen - 2
		const newPointsOnEdge = Math.floor(pointsToAdd / edges)
		const bufferWithPointsEveryEdge = bufferLen + newPointsOnEdge * lastPoint
		let remainingPoints = (finalBufferLength - bufferWithPointsEveryEdge) / 2
		const edgeRemainingIndex = Math.round(edges / remainingPoints)
		const result = new Float32Array(finalBufferLength)

		for (let i = 0, edgeIndex = 0, r = 0; i < lastPoint; i += 2, edgeIndex++, r += 2) {
			const ax = buffer[i]
			const ay = buffer[i + 1]
			const bx = buffer[i + 2]
			const by = buffer[i + 3]
			result[r] = ax
			result[r + 1] = ay

			const addReminingPoints = remainingPoints > 0 && (edgeIndex % edgeRemainingIndex === 0 || i === lastPoint - 2)
			const currentPointsOnEdge = newPointsOnEdge + (addReminingPoints ? 1 : 0)
			const newPointOffset = 1 / (currentPointsOnEdge + 1)

			for (let h = 0; h < currentPointsOnEdge; h++, r += 2) {
				const o = newPointOffset * (h + 1)
				result[r + 2] = (1 - o) * ax + o * bx
				result[r + 3] = (1 - o) * ay + o * by
			}
			if (addReminingPoints) {
				remainingPoints--
			}
		}
		result[finalBufferLength - 2] = buffer[bufferLen - 2]
		result[finalBufferLength - 1] = buffer[bufferLen - 1]

		return result
	}

	const result = new Float32Array(finalBufferLength)

	for (let i = 0; i < finalBufferLength; i += 2) {
		result[i] = buffer[i % bufferLen]
		result[i + 1] = buffer[(i + 1) % bufferLen]
	}

	return result
}

/**
 * Leads two buffers to have the same number of points
 *
 * @category Utilities.Buffer interpolation
 * @param from
 * @param to
 * @returns
 */
export function prepareBufferForInterpolation(from: Float32Array, to: Float32Array): [Float32Array, Float32Array] {
	const fromBufferLength = from.length
	const toBufferLength = to.length

	if (fromBufferLength === toBufferLength) {
		return [from, to]
	}

	// const maxBufferLength = fromBufferLength > toBufferLength ? fromBufferLength : toBufferLength
	const difference = Math.abs(fromBufferLength - toBufferLength)
	// const minBufferLength = maxBufferLength - difference

	/////

	const b = fromBufferLength < toBufferLength ? to : from
	const t = fromBufferLength < toBufferLength ? from : to

	const a = distributePointsInBuffer(t, Math.floor(difference / 2))

	// a[maxBufferLength - 2] = t[minBufferLength - 2]
	// a[maxBufferLength - 1] = t[minBufferLength - 1]

	return fromBufferLength > toBufferLength ? [b, a] : [a, b]
}

/**
 * Interpolate two buffer
 *
 * @category Utilities.Buffer interpolation
 * @param from
 * @param to
 * @param offset
 * @returns
 */
export function interpolate(
	from: Float32Array,
	to: Float32Array,
	initialOffset: number | Array<number> = 0.5
): Float32Array {
	const [a, b] = prepareBufferForInterpolation(from, to)
	const maxBufferLength = Math.max(a.length, b.length)
	const offset = typeof initialOffset === 'number' ? [initialOffset] : initialOffset
	const maxPoints = maxBufferLength / 2

	if (offset.length !== maxPoints) {
		const tl = offset.length
		for (let i = 0; i < maxPoints; i++) {
			offset[i] = offset[i % tl]
		}
	}

	////

	const result = new Float32Array(maxBufferLength)

	for (let i = 0, off = 0; i < maxBufferLength; i += 2, off++) {
		result[i] = (1 - offset[off]) * a[i] + offset[off] * b[i]
		result[i + 1] = (1 - offset[off]) * a[i + 1] + offset[off] * b[i + 1]
	}

	return result
}
